### PR TITLE
JSON-RPC: Extend jamulusserver/getClients method

### PR DESCRIPTION
**Pull request by Rob-NY**
**Short description of changes**

Expands jamulusserver/getClients method to include all elements of a user profile as well as the number of active client connections.  Changes were made so they are backward compatible with the prior message format.

CHANGELOG:  jamulusserver/getClients method expanded to include all elements of a user profile as well as the number of active client connections.  

**Context: Fixes an issue?**

Addresses issue #2488 and comments in #2890 

**Does this change need documentation? What needs to be documented and how?**

JSON-RPC documentation was updated via python tool and is included in this PR.

**Status of this Pull Request**

**What is missing until this pull request can be merged?**

Compile tested on Linux only.

## Checklist

-  [X] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [X] I tested my code and it does what I want
-  [X] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
-  [X] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [X] I've filled all the content above

### CODE REVIEW

File: src/util.h
Comment (pljones) [inline src/util.h:680]: Why is  this bound to `"CServerDlg"`?  Why not just re-use the existing look-up method?

Comment (pljones) [reply src/util.h:680]: As a reminder - **_NO_** functionality - and _this_ counts as functionality - should exist "in the UI".  So explicitly making a second copy of something that should not exist really is bad...

Comment (pljones) [reply src/util.h:680]: Ideally `server.cpp` would have a method to do the lookup that delegated to a shared utility method also used by `client.cpp`... although, in both cases, likely delegated via `channel.cpp`.

Comment (Rob-NY) [reply src/util.h:680]: > Why is this bound to `"CServerDlg"`? Why not just re-use the existing look-up method?

Perhaps I missed where this exists today... I'm happy to change it.

File: src/serverrpc.cpp
Comment (Rob-NY) [reply src/serverrpc.cpp:113]: I believe vecHostAdddresses.Size() returns the number of slots, not the number of active connections.

File: src/util.h
Comment (Rob-NY) [reply src/util.h:680]: > As a reminder - **_NO_** functionality - and _this_ counts as functionality - should exist "in the UI". So explicitly making a second copy of something that should not exist really is bad...

What is in the UI that you object to?   The change to CServerDlg is required because a method prototype was changed in CServer that is called **from** CServerDlg.   

And, again, if I missed an existing method for converting skill number to skill literal I will happily refactor -- but I didn't find one (perhaps I missed it?)

Comment (pljones) [reply src/util.h:680]: UI is any user interface component.  That means the graphical user interface, the RPC interface and MIDI controls for the Jamulus core code.

Comment (pljones) [reply src/util.h:680]: > And, again, if I missed an existing method for converting skill number to skill literal I will happily refactor -- but I didn't find one (perhaps I missed it?)

Like I say, there's a _lot_ of code in the UI that shouldn't be there.  Currently the mapping is done _literally_ in the UI: src/clientsettingsdlg.cpp lines 611 to 629.

I'm happy adding the method here, as you've done.  The first argument in the `QCoreApplication::translate` call (`"CServerDlg"` or whatever) doesn't seem used - I can't see anything in the *.ts files referencing it.  The *.ts files _do_ reference the containing file (`util.h` here).  I'd rather it just said "util", and we moved away from trying to guess where utility methods got called from...

Comment (Rob-NY) [reply src/util.h:680]: Sorry, but I do not know how to respond to this.  I don't see where there is "functionality" in the RPC code that is any different from the first release of RPC.  It merely calls the relevant methods elsewhere (just as the first version of RPC did).   And if you're only referring to skill code translation, even that is in util.* where other core code is -- not in the RPC code.   So I'm sorry @pljones, I'm not following your comments enough to know how to effect a change.

File: src/server.cpp
Comment (pljones) [inline src/server.cpp:1483]: This isn't used by `server.cpp` then?  Why set it here?

File: src/util.h
Comment (pljones) [reply src/util.h:680]: `src/clientsettingsdlg.cpp` isn't the RPC code.  You're duplicating code here that's literally in the graphical user interface.  It shouldn't be.  That's what I'm complaining about.  The trigger for my complaint was the mention of the Service Graphical User Interface.  Which is irrelevant to what you're doing and hence _wrong_ and needs changing.

I've said that this is the _right_ place to do it.  Ideally the code in `src/clientsettingsdlg.cpp` would get fixed to call this method, too, and the translation would only happen in one place, consistently.

Comment (hoffie) [reply src/util.h:680]: Heh, I was confused about the subject of complaint as well as the primary indicator for the comment thread is... well, the new code, which was *not* the subject. ;)

File: src/serverrpc.cpp
Comment (pljones) [reply src/serverrpc.cpp:113]: But I see this at line 86:
```
// fill list with connected clients
for ( int i = 0; i < iNumChannels; i++ )
```

Comment (pljones) [reply src/serverrpc.cpp:113]: Oh and at line 60
```
/// @result {array} result.clients - The list of connected clients.
```

Comment (pljones) [reply src/serverrpc.cpp:113]: Rather than call GetNumberOfConnectedClients, having actively filtered for the connected clients, just keep your own local count and use it.  You know you'll get the right number, that way.  Nothing will have changed between one call and the next.

File: src/serverdlg.cpp
Comment (pljones) [inline src/serverdlg.cpp:693]: What does the Graphical user interface want to do this for?

File: src/util.h
Comment (pljones) [inline src/util.h:697]: There shouldn't be a default - all the options should be covered and "None" should be returned for SL_NOT_SET

File: src/serverdlg.cpp
Comment (Rob-NY) [reply src/serverdlg.cpp:693]: While not my code, the return values of GetConCliParam are used to populate the connection box at the top of the server dialog.  This method call was already here;  I just added the additional parameter.   This _exact_ discussion was had in another thread that I cannot find now -- but will keep looking.

Comment (pljones) [reply src/serverdlg.cpp:693]: I'm not keen on unused variables (neither is the IDE).  If there are two different use cases, there should be two supporting methods.  One might call the other with a dummy parameter - or pass a dummy value here and ignore it in `GetConCliParam`.  At the moment, extra work is being done unnecessarily.

File: src/util.h
Comment (Rob-NY) [reply src/util.h:680]: > The trigger for my complaint was the mention of the Server Graphical User Interface. Which is irrelevant to what you're doing and hence _wrong_ and needs changing.

we covered this here: https://github.com/jamulussoftware/jamulus/issues/2488#issuecomment-1065865193

The challenge with GetConCliParam is that it entirely uses pass-by-reference parameters.  I’m not sure it is reasonable to expect every caller to this method to have to use every reference returned.   Before this PR, the only place this method was called was from the server dialog.   I thought it was more efficient to modify and re-use this method instead of creating an entirely new one for RPC.   In fact, the server dialog might benefit now from the additional channel info that was never available to it before; I’ll look into this.

File: src/serverrpc.cpp
Comment (Rob-NY) [reply src/serverrpc.cpp:113]: iNumChannels is the size of vecHostAddresses which 150 with the standard compile -- representing the max size of the all these vectors which is hard-coded as 15 (global.h:209)

It is true that I could count the loop through, but with large # of active connections would represent 'x' number of operations as opposed to a single call to a method that already knows the answer.

File: src/server.cpp
Comment (Rob-NY) [reply src/server.cpp:1483]: This method was **never called** by CServer and is not used by CServer.  Prior to this PR, it was only called by server dialog which, as discussed below, now will also has access to the chanInfo that it didn't have before (albeit not yet used).  See discussion #2488

File: src/util.h
Comment (Rob-NY) [reply src/util.h:697]: I will update with an SL_NOT_SET case -- but I believe having a default is prudent;  perhaps "Not defined" (untranslated).

Comment (Rob-NY) [reply src/util.h:697]: It is worth noting from a semantics perspective that saying "Skill -> None" can mean both "having no skill" and "having not made a selection"  

image

I'm not fighting the suggestion, just making the point that an empty string is more reflective of "not making a choice" than the word "None" is.

File: src/serverrpc.cpp
Comment (pljones) [reply src/serverrpc.cpp:113]: i++ no more than 150 times is probably a _lot_ cheaper than the method call (allocate memory for the stack, put return address on stack, allocate memory for any returned values, make the call, copy local variable to stack, return from method, copy stack to local variable, free stack... depends how expensive getting the memory turns out to be...).

But imagine now you're in a separate thread.  And the JSONRPC code _should_ be running in a separate thread from the real-time server code.  You're now not only doing that in process call, you're doing a cross-thread context switch.

File: src/util.h
Comment (pljones) [reply src/util.h:697]: It's about consistency.  If the "None" in the English UI is changed to "Not specified", for example, the JSONRPC response should change, if you're expecting to get the human-readable value.

**Comment by ann0see** [comment]
I'd like to have this in the next release...

**Comment by Rob-NY** [comment]
> I'd like to have this in the next release...

Me too.  I've fixed what I believe are the issues raised (as opposed to philosophical questions), squashed and pushed.  Once this is merged I will resubmit the depended PR.

File: src/serverrpc.cpp
Comment (hoffie) [inline src/serverrpc.cpp:105]: Do we have a specific use case for exposing the numeric country code?
Unless there is a very good reason, we should not expose it here. IIRC the country code is Qt-native here. This means, the interpretation of this code would be influenced by whether a server is running a Qt5 or a Qt6 build.

If we do want to expose something besides the country name, we should expose the ISO code, I think.

File: src/util.h
Comment (hoffie) [reply src/util.h:706]: @pljones Sorry to join this conversation that late, but do we really want to expose the translated string here? Wouldn't this mean that the API client would need to handle different server languages...?

**Review by hoffie — COMMENTED** [review]
Thanks for the PR and sorry for not commenting earlier. Looks good to me in general despite two questions (see inline comments).

File: src/util.h
Comment (pljones) [reply src/util.h:706]: If I'm writing a web site module that interacts over the JSONRPC interface, I don't want to have to worry about the API response getting translated because of the country Qt thinks the software is running in.  The API responses should always be the same _value_ for the same _meaning_.  How the web site then displays that in _its_ user interface is a different issue and outside the scope of JSONRPC.

**Comment by Rob-NY** [comment]
> I'd like to have this in the next release...

@hoffie - is there anything I still need to do to move this along?  I'm asking because I'd like to resubmit the dependent PR which builds upon this basic one as a candidate as well.  I believe I've addressed all the refactor requests and questions.

File: src/util.h
Comment (pljones) [reply src/util.h:706]: @Rob-NY This is still outstanding.

File: src/serverrpc.cpp
Comment (ann0see) [reply src/serverrpc.cpp:105]: @Rob-NY This is still outstanding I think

Comment (ann0see) [reply src/serverrpc.cpp:105]: @Rob-NY ?

**Comment by pljones** [comment]
@Rob-NY there are a number of unanswered requests.

**Comment by Rob-NY** [comment]
> @Rob-NY there are a number of unanswered requests.

ISO Country code standardization:  I have no idea how to approach this at this time, so I asked to pass on this for the time being - no response.

In November, @hoffie asked you if we really wanted to expose the translated strings in some areas of this api - no response.

I made the changes for the "Skillset: NONE" vs. "not set" request as well as all the others raised here.   If I missed something, please point it out to me and I'll address right away.

**Comment by ann0see** [comment]
Ok. So then there was a misunderstanding on both sides about the state of this PR. 

I‘ll look where I can help if I have time.

File: src/serverrpc.cpp
Comment (ann0see) [reply src/serverrpc.cpp:105]: > If we do want to expose something besides the country name, we should expose the ISO code, I think.


This would probably be https://doc.qt.io/qt-6.2/qlocale.html#codeToScript or https://doc.qt.io/qt-6.2/qlocale.html#bcp47Name or https://doc.qt.io/qt-5/qlocale.html#name then.

File: src/util.h
Comment (ann0see) [reply src/util.h:706]: Why can't we just return the int? If the API is well documented, the number should be sufficient.

File: docs/JSON-RPC.md
Comment (ann0see) [inline docs/JSON-RPC.md:246]: Could someone please check if this returns the correct value?

I set the city to "...." but JSON-RPC just gives me:

```
{"id":1,"jsonrpc":"2.0","result":{"clients":[{"address":"127.0.0.1:22217","channels":2,"city":"","countryCode":0,"countryName":"Default","id":0,"instrumentCode":0,"instrumentName":"Nichts","jitterBufferSize":2,"name":"eto22","skillLevelCode":0,"skillLevelName":"Keins"}],"connections":1}}
```

Comment (ann0see) [reply docs/JSON-RPC.md:246]: @Rob-NY could you please test this on your machine?

File: src/util.h
Comment (ann0see) [inline src/util.h:707]:
```suggestion
    return iSkill;
```

I'd rather remove this.

File: src/serverrpc.cpp
Comment (ann0see) [reply src/serverrpc.cpp:105]: After further thinking, probably the countryName is sufficient.

File: src/serverrpc.cpp
Comment (ann0see) [inline src/serverrpc.cpp:67]: I think we could argue to also remove this, as it's a Jamulus protocol ID. But to get this PR going, I'd accept it for now.

File: src/serverrpc.cpp
Comment (ann0see) [reply src/serverrpc.cpp:105]: The issue here is the QT5/QT6 change and inconsistency.

**Review by ann0see — CHANGES_REQUESTED** [review]
Ok. See https://github.com/ann0see/jamulus/commit/f3d6bea9a89eb71a081a17a5f7ece7111ad5300f 
I hope we can agree on removing these methods as:
1. The country code is inconsistent between QT5 and QT6 (we already expose the country name which is not optimal but probably ok)
2. The skill level is translated and thus inconsistent. Alternatively, we'd find a way to only use the English translation. However, I think using just IDs is fine in this case, as the Jamulus protocol specifies the meaning. I believe that if we remove this - to stay consistent we should also remove the text translations of other translated names.

File: docs/JSON-RPC.md
Comment (ann0see) [reply docs/JSON-RPC.md:246]: Also I believe that instrumentName is not a number but rather a String. As mentioned below, I think - as this is translated - we should think of removing it.

**Comment by ann0see** [comment]
@pljones @Rob-NY would you be ok with my proposed changes?

File: src/util.h
Comment (pljones) [reply src/util.h:706]: Still outstanding.

File: src/util.h
Comment (pljones) [reply src/util.h:707]: Please, this is better.

**Comment by Rob-NY** [comment]
> @pljones @Rob-NY would you be ok with my proposed changes?

No issue.

**Comment by ann0see** [comment]
@Rob-NY @pljones pushed my commit. Waiting on further review now.

**Comment by ann0see** [comment]
Ok. Rebased. Will do a new local test now.

**Review by ann0see — APPROVED** [review]
Local testing says this is ok. still, we could remove the instrument name - as it is translated and thus not reliable. Qt doesn't seem to support any short name for countries, so the country name is probably the best we can get.

**Comment by ann0see** [comment]
@pljones (and maybe @dtinth) Could you please review this (especially from a C++ dev viewpoint)? I'd like to have this in. The legacy and iOS build failures are unrelated.

**Review by dtinth — APPROVED** [review]
Looks good to me too.

`vecChanInfo` in `CServerDlg` seems unused but that's fine I guess.

File: docs/JSON-RPC.md
Comment (pljones) [inline docs/JSON-RPC.md:245]: `instrumentName | number`?  That sounds strange.  What is it meant to represent?  The position in the enum?  In which case it's redundant.

**Comment by pljones** [comment]
I'd like a new issue raised with the changes requested here but not made.  Once that's done, I'm happy to approve it.

File: docs/JSON-RPC.md
Comment (ann0see) [reply docs/JSON-RPC.md:245]: >  we could remove the instrument name

I think this is wrongly documented and should be a string (but I still believe it should be removed.)

**Comment by ann0see** [comment]
@pljones 

For fixable stuff, I'd only have the removal of translated content (= instrument name). 
I believe there is more, but I think others have raised these issues:
* https://github.com/jamulussoftware/jamulus/pull/2918#discussion_r997291702
* https://github.com/jamulussoftware/jamulus/pull/2918#discussion_r997459985
* https://github.com/jamulussoftware/jamulus/pull/2918#pullrequestreview-1419847382
(feel free to add your comments to this list and then open an issue - it seems that most open stuff is related to design issues in the original code you found.)

**Comment by pljones** [comment]
> @pljones
> 
> For fixable stuff, I'd only have the removal of translated content (= instrument name). I believe there is more, but I think others have raised these issues:

Yes, it's probably mostly the use of strings where the number would be better that should be fixed (at some point):
https://github.com/jamulussoftware/jamulus/pull/2918#discussion_r1191588195

**Comment by ann0see** [comment]
Ok. I've now removed the instrument name too. 
@pljones could you please go through the comments (especially your comments) and open the issue? I think you have more context.

**Review by pljones — APPROVED** [review]
I'll approve anyway on the basis nothing here is breaking as far as I can see.  The `countryName` issue is complicated because of Qt 5/6 rather than anything else, as far as I remember and I wasn't involved much in the solution there.

**Pull request merged**
